src/translate/pddl_parser/parsing_functions.py

```diff
@@ -313,186 +547,168 @@ def parse_task(domain_pddl, task_pddl):
+            yield domain_line[1]
+
+        requirements_opt = next(iterator)
+        context.check_expected_block(requirements_opt, [":requirements", ":objects", ":init"])
+        if requirements_opt[0] == ":requirements":
+            requirements = requirements_opt[1:]
+            objects_opt = next(iterator)
         else:
-            atom = pddl.Atom(fact[0], fact[1:])
-            check_atom_consistency(atom, initial_true, initial_false)
-            initial_true.add(atom)
-    initial.extend(initial_true)
-    yield initial
-
-    goal = next(iterator)
-    assert goal[0] == ":goal" and len(goal) == 2
-    yield parse_condition(goal[1], type_dict, predicate_dict)
-
-    use_metric = False
-    for entry in iterator:
-        if entry[0] == ":metric":
-            if entry[1] == "minimize" and entry[2][0] == "total-cost":
-                use_metric = True
-            else:
-                assert False, "Unknown metric."
-    yield use_metric
+            requirements = []
+            objects_opt = requirements_opt
+        yield parse_requirements(context, requirements)
+
+        context.check_expected_block(objects_opt, [":objects", ":init"])
+        if objects_opt[0] == ":objects":
+            with context.layer("Parsing objects"):
+                yield parse_typed_list(context, objects_opt[1:])
+            init = next(iterator)
+        else:
+            yield []
+            init = objects_opt
+
+        yield parse_init(context, init)
+
+        goal = next(iterator)
+        context.check_expected_block(goal, [":goal"])
+        with context.layer("Parsing goal"):
+            syntax = "'(:goal GOAL)'"
+            if len(goal) != 2:
+                context.error(f"Expecting {syntax}.")
+            context.check_list(goal[1], min_size=1, block_name="Goal condition",
+                               finalmsg="Expecting non-empty goal condition.")
+            if not goal[1]:
+                context.error(f"Expecting non-empty goal {syntax}")
+            yield parse_condition(context, goal[1], type_dict, predicate_dict)
+
+        use_metric = False
+        for entry in iterator:
+            context.check_expected_block(entry, [":metric"])
+            with context.layer("Parsing metric"):
+                if len(entry) != 3:
+                    context.error("Expecting '(:metric minimize (total-cost))'")
+                if entry[1] == "minimize" and entry[2][0] == "total-cost":
+                    use_metric = True
+                else:
+                    context.error("Expecting '(:metric minimize (total-cost))'")
+        yield use_metric
 
-    for entry in iterator:
-        assert False, entry
+        for _ in iterator:
+            context.error("No blocks expected after goal and metric.")
 
 
-def check_atom_consistency(atom, same_truth_value, other_truth_value, atom_is_true=True):
+def check_atom_consistency(context, atom, same_truth_value, other_truth_value, atom_is_true=True):
```

Comment:
Since you moved `check_for_duplicates` to the Context class we could also move this one. But I don't have a strong opinion, thematically it fits less to Context than the other one...

Reply:
I think you  are right, there are some similarities between `check_for_duplicates` and `check_atom_consistency`. The functions in `Context` perform checks on unparsed tokens. None of the functions named before does this. Thus, I pulled `check_for_duplicates` out again. This also shrinks the diff.